src/Chainweb/Version/Utils.hs

```diff
@@ -0,0 +1,434 @@
+{-# INLINE lastGraphChange #-}
+
+nextGraphChange
+    :: HasCallStack
+    => HasChainwebVersion v
+    => v
+    -> BlockHeight
+    -> BlockHeight
+nextGraphChange v h = fst . fromJuste . M.lookupGT h $ chainGraphs v
+{-# INLINE nextGraphChange #-}
+
+chainCountAt
+    :: HasCallStack
+    => HasChainwebVersion v
+    => v
+    -> BlockHeight
+    -> Natural
+chainCountAt v h = order $ atHeight h $ chainGraphs v
+{-# INLINE chainCountAt #-}
+
+degreeAt
+    :: HasCallStack
+    => HasChainwebVersion v
+    => v
+    -> BlockHeight
+    -> Natural
+degreeAt v h = degree $ atHeight h $ chainGraphs v
+{-# INLINE degreeAt #-}
+
+diameterAt
+    :: HasCallStack
+    => HasChainwebVersion v
+    => v
+    -> BlockHeight
+    -> Natural
+diameterAt v h = diameter $ atHeight h $ chainGraphs v
+{-# INLINE diameterAt #-}
+
+chainIdsAt
+    :: HasCallStack
+    => HasChainwebVersion v
+    => v
+    -> BlockHeight
+    -> HS.HashSet ChainId
+chainIdsAt v h = graphChainIds $ atHeight h $ chainGraphs v
+{-# INLINE chainIdsAt #-}
+
+-- | Uniformily get a random ChainId at the top of the current chainweb
+--
+randomChainId :: HasChainwebVersion v => v -> IO ChainId
+randomChainId v = randomChainIdAt v maxBound
+{-# INLINE randomChainId #-}
+
+-- | Uniformily get a random ChainId at the given height of the chainweb
+--
+randomChainIdAt :: HasChainwebVersion v => v -> BlockHeight -> IO ChainId
+randomChainIdAt v h = (!!) (toList cs) <$> randomRIO (0, length cs - 1)
+  where
+    cs = chainIdsAt v h
+{-# INLINE randomChainIdAt #-}
+
+-- | Sometimes, in particular for testing and examples, some fixed chain id is
+-- needed, but it doesn't matter which one. This function provides some valid
+-- chain ids for the top of the current chainweb.
+--
+someChainId :: HasCallStack => HasChainwebVersion v => v -> ChainId
+someChainId v = someChainIdAt v maxBound
+{-# INLINE someChainId #-}
+
+-- | Sometimes, in particular for testing and examples, some fixed chain id is
+-- needed, but it doesn't matter which one. This function provides some valid
+-- chain ids for the chainweb at the given height.
+--
+someChainIdAt :: HasCallStack => HasChainwebVersion v => v -> BlockHeight -> ChainId
+someChainIdAt v h = head . toList $ chainIdsAt v h
+    -- 'head' is guaranteed to succeed because the empty graph isn't a valid chain
+    -- graph.
+{-# INLINE someChainIdAt #-}
+
+isGraphChange :: HasChainwebVersion v => v -> BlockHeight -> Bool
+isGraphChange v h = M.member h (chainGraphs v)
+{-# INLINE isGraphChange #-}
+
+-- -------------------------------------------------------------------------- --
+-- Block Count
+
+-- | The numbers of blocks on a chain up to the given height
+--
+-- Precondition: h > genesisHeight
+--
+blockCountAt
+    :: HasCallStack
+    => HasChainwebVersion v
+    => HasChainId cid
+    => v
+    -> cid
+    -> BlockHeight
+    -> Natural
+blockCountAt v cid h
+    | h < gh = error $ "Chainweb.Version.Utils.blockCount: requested block height " <> sshow h
```

Comment:
Suggest one of two things:
- return 0 for < genesis heights, keeps function pure 
- make function safe (return `Maybe Natural`)
Depending on usage site requirements and frequency of use. ie if this is just for difficulty and we expect this to never happen, a `0` is acceptable and cleaner.

Reply:
Yes, 0 seems like the correct result.

On the other hand, I sometimes prefer a clean crash with an informative error message over computations continuing under possibly unexpected assumptions. At least during development and debugging, it was very helpful having those errors, that would trigger when some invariant was violated.